blog_post_collaboration/spark_streaming_dataset.md

```diff
@@ -0,0 +1,322 @@
+    
+    # Directory that holds configuration.
+    # CONF_SOURCE = "conf"
+    
+    # Class that manages how configuration is loaded.
+    # CONFIG_LOADER_CLASS = ConfigLoader
+    # Keyword arguments to pass to the `CONFIG_LOADER_CLASS` constructor.
+    # CONFIG_LOADER_ARGS = {
+    #       "config_patterns": {
+    #           "spark" : ["spark*/"],
+    #           "parameters": ["parameters*", "parameters*/**", "**/parameters*"],
+    #       }
+    # }
+    
+    # Class that manages the Data Catalog.
+    # from kedro.io import DataCatalog
+    # DATA_CATALOG_CLASS = DataCatalog
+```
+
+
+## How to set up your Kedro project to read data from streaming sources
+
+Once you have set up your project, you can use the new [Kedro Spark streaming dataset](https://docs.kedro.org/en/stable/kedro_datasets.spark.SparkStreamingDataSet.html). You need to configure the data catalog, in `conf/base/catalog.yml` as follows:
+
+### Reading a streaming JSON file
+
+```yaml
+raw_json:
+  type: spark.SparkStreamingDataSet
+  filepath: data/01_raw/stream/inventory/
+  file_format: json
+```
+
+Additional options can be configured via the `load_args` key.
+
+```yaml
+int.new_inventory:
+   type: spark.SparkStreamingDataSet
+   filepath: data/02_intermediate/inventory/
+   file_format: csv
+   load_args:
+     header: True
+```
+
+## How to set up your Kedro project to write data to streaming sinks
+
+All the additional arguments can be kept under the `save_args` key:
+
+```yaml
+processed.sensor:
+   type: spark.SparkStreamingDataSet
+   file_format: csv
+   filepath: data/03_primary/processed_sensor/
+   save_args:
+     output_mode: append
+     checkpoint: data/04_checkpoint/processed_sensor
+     header: True
+```
+
+Note that when you use the Kafka format, the respective packages should be added to the `spark.yml` configuration as follows:
+
+```yaml
+spark.jars.packages: org.apache.spark:spark-sql-kafka-0-10_2.12:3.3.1 
+```
+    
+## Design considerations
+### Pipeline design
+
+In order to benefit from Spark's internal query optimisation, we recommend that any interim datasets are stored as memory datasets.
+
+All streams start at the same time, so any nodes that have a dependency on another node that writes to a file sink (i.e. the input to that node is the output of another node) will fail on the first run. This is because there are no files in the file sink for the stream to process when it starts. 
+
+We recommended that you either keep intermediate datasets in memory or split out the processing into two pipelines and start by triggering the first pipeline to build up some initial history.
+
+### Feature creation
+
+Be aware that windowing operations only allow windowing on time columns.
+
+Watermarks must be defined for joins. Only certain types of joins are allowed, and these depend on the file types (stream-stream, stream-static) which makes joining of multiple tables a little complex at times. For further information or advice......TO DO.
```

Comment:
@harisqb @kuriantom369 @tingtingQB Please could you help me complete this here -- it may be simply that they need to come to the Slack organisation to ask a question, but I think if we are saying that something is a little complex, it's necessary to say how to find out more to resolve any issues caused by complexity.

Reply:
Mainly there is a good docs in the pyspark website, the link is mentioned below. It gives a clear idea about various join types, watermarking, and considerations. 

https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#join-operations